packages/angular/ssr/src/i18n.ts

```diff
@@ -43,3 +43,107 @@ export function getPotentialLocaleIdFromUrl(url: URL, basePath: string): string
+ */
+interface LocalePreference {
+  /** The locale (e.g., 'en-US', 'fr') */
+  locale: string;
+
+  /** The quality value indicating the preference */
+  quality?: number;
+}
+
+/**
+ * Parses the `Accept-Language` header and returns a list of locale preferences with their respective quality values.
+ *
+ * @param header - The value of the `Accept-Language` header, typically a comma-separated list of locales with optional quality values.
+ * @returns A list of `LocalePreference` objects representing the parsed locales and their associated quality values.
+ *
+ * @example
+ * ```js
+ * parseLanguageHeader('en-US;q=0.8,fr-FR;q=0.9')
+ * // returns [{ locale: 'en-US', quality: 0.8 }, { locale: 'fr-FR', quality: 0.9 }]
+ * ```
+ */
+function parseLanguageHeader(header: string): LocalePreference[] {
+  if (header === '*') {
+    return [{ locale: '*', quality: undefined }];
+  }
+
+  const localesList: LocalePreference[] = [];
+  const items = header.split(',');
+
+  for (const item of items) {
+    const [locale, qualityValue] = item
+      .trim()
+      .split(';', 2)
+      .map((v) => v.trim());
+
+    const quality = qualityValue?.startsWith('q=') ? parseFloat(qualityValue.slice(2)) : undefined;
+
+    localesList.push({
+      locale,
+      quality: quality !== undefined && quality <= 1 ? quality : undefined,
+    });
+  }
+
+  return localesList;
+}
+
+/**
+ * Gets the preferred locale based on the highest quality value from the provided header and the set of available locales.
+ * If no exact match is found, it tries to find the closest match based on language prefixes.
+ *
+ * @param header - The `Accept-Language` header string to parse and evaluate.
+ * @param availableLocales - A readonly set of available locales that the user can choose from.
+ * @returns The best matching locale.
+ *
+ * @example
+ * ```js
+ * getPreferredLocale('en-US;q=0.8,fr-FR;q=0.9', new Set(['en-US', 'fr-FR', 'de-DE']))
+ * // returns 'fr-FR'
+ * ```
+ */
+export function getPreferredLocale(header: string, availableLocales: ReadonlySet<string>): string {
+  const parsedLocales = parseLanguageHeader(header);
+  if (
+    parsedLocales.length === 0 ||
+    (parsedLocales.length === 1 && parsedLocales[0].locale === '*')
+  ) {
+    return availableLocales.values().next().value as string;
+  }
+
+  let bestMatch: string | undefined;
+  let bestQuality = 0;
+
+  // Find the best exact match based on quality
+  for (const { locale, quality } of parsedLocales) {
+    const currentQuality = quality ?? 1;
+    if (availableLocales.has(locale) && currentQuality > bestQuality) {
+      bestMatch = locale;
+      bestQuality = currentQuality;
+    }
+  }
+
+  if (bestMatch) {
+    return bestMatch;
+  }
+
+  // If no exact match, fallback to closest matches
+  bestQuality = 0;
+  for (const { locale, quality } of parsedLocales) {
+    const currentQuality = quality ?? 1;
+    const languagePrefix = locale.split('-', 1)[0];
```

Comment:
Those are not used in the header. See: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Language#directives

> A language tag (which is sometimes referred to as a "locale identifier"). This consists of a 2-3 letter base language tag that indicates a language, optionally followed by additional subtags separated by -. The most common extra information is the country or region variant (like en-US or fr-CA) or the type of alphabet to use (like sr-Latn). Other variants, like the type of orthography (de-DE-1996), are usually not used in the context of this header.

Reply:
> usually not used

Is that a hard requirement of the spec? Does it limit to just a tag and a subtag?

I agree it's unlikely, I'm just aiming for compatibility with the spec here.

Reply:
The spec mentioned that subtags meaning is not defined and it will result in unpredictable behaviour

https://datatracker.ietf.org/doc/html/rfc4647

> The value and semantic meaning of private-use tags and of the subtags
   used within such a language tag are not defined.  Matching private-
   use tags using language ranges or extended language ranges can result
   in unpredictable content being returned.

Reply:
Hmm, that does seem vague enough that we can mostly do what we want here. I'm ok to limit to only 2 locale segments for now if that makes the implementation easier. Should we consider throwing an error if an application defines a supported locale with 3 or more segments? Not sure if those would ever be matched.

I do kind of suspect that eventually someone will file a bug to support 3+ segments, but I'm fine to wait for that use case if/when it comes.